### PR TITLE
Fix schema mismatch, race conditions, and error handling issues

### PR DESCRIPTION
**Pull request by leynos**
This commit addresses multiple code review findings:

Database and Migration Fixes:
- Fix v16_context.rs test fixture to include V13 columns (max_tokens, total_tokens_used)
- Add migration race condition protection by re-checking _migrations under lock
- Add limit > 0 validation in list_conversation_messages_paginated
- Replace unchecked cast with i64::from for action.sequence
- Fix decode error handling for dedup_window_secs and tokens_used in routines
- Add affected row checks to UPDATE statements in routine_runs
- Fix usize truncation on 32-bit targets in agent_job_summary

Error Handling Improvements:
- Convert panics to Result propagation in singletons.rs test helpers
- Remove fire-and-forget persistence in job.rs, use synchronous awaits

Performance and Correctness:
- Optimize has_settings query to use EXISTS instead of COUNT(*)
- Use cloned() instead of ToString::to_string for key cloning
- Use upsert instead of read-modify-replace in save_default_settings
- Persist job mode after restart in restart.rs

Build and Test:
- Add conditional compilation for backend-specific test helpers
- Add comment explaining description column reuse for credential grants

Co-authored-by: factory-droid[bot] <138933559+factory-droid[bot]@users.noreply.github.com>

## Summary by Sourcery

Harden database migrations, persistence, and job handling logic to fix race conditions, schema mismatches, and error reporting, while tightening pagination and numeric handling across backends.

Bug Fixes:
- Prevent duplicate or skipped libSQL migrations by re-checking _migrations under transaction lock before applying V16 context-field changes.
- Correct the V16 libSQL migration test fixture to include columns added in earlier migrations and mark older migrations as applied to match real schema evolution.
- Return NotFound errors when updating or linking routine_runs that affect no rows instead of silently succeeding.
- Validate that conversation message pagination limits are greater than zero to avoid invalid or overflow-prone queries.
- Fix decoding of optional numeric routine fields by using typed Option<i64> accessors and surfacing clear serialization errors.
- Avoid usize truncation on 32-bit targets when building agent job summaries by fallibly converting from i64.
- Use safe integer conversion for job action sequence values to prevent overflow and undefined behavior.

Enhancements:
- Refactor libSQL migration code to separate state checking, recording, and transaction finalization for clearer error handling.
- Propagate database errors from singleton test helpers instead of panicking, and assert success explicitly in tests.
- Change default settings persistence to upsert individual keys to avoid races with other writers.
- Optimize settings existence checks to use EXISTS instead of COUNT(*) for improved performance.
- Use cloned keys instead of to_string when deleting obsolete settings for minor allocation efficiency gains.
- Persist sandbox job records and ClaudeCode job modes synchronously during creation and restart to guarantee durability before container startup.
- Document the reuse of the sandbox job description column for serialized credential grants to clarify schema semantics.
- Add conditional compilation guards for backend-specific trybuild database forwarder helpers to match enabled features.

Tests:
- Adjust migration tests to more accurately simulate real-world migration ordering and schema contents for V16 context migrations.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Large Method
src/db/libsql_migrations/v16_context.rs: apply_agent_jobs_context_fields_migration

### What lead to degradation?
apply_agent_jobs_context_fields_migration has 78 lines, threshold = 70

### Why does this problem occur?
Overly long functions make the code harder to read. The recommended maximum function length for the Rust language is 70 lines of code. Severity: Brain Method - Complex Method - Long Method.

### How to fix it?
We recommend to be careful here -- just splitting long functions don't necessarily make the code easier to read. Instead, look for natural chunks inside the functions that expresses a specific task or concern. Often, such concerns are indicated by a Code Comment followed by an if-statement. Use the [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html) refactoring to encapsulate that concern.

File: src/tools/builtin/job.rs [hunk 1]
```diff
@@ -362,21 +351,23 @@ impl CreateJobTool {
...
+                store
+                    .update_sandbox_job_mode(job_id, crate::db::SandboxMode::ClaudeCode)
                     .await
-                {
-                    tracing::warn!(job_id = %job_id_copy, "Failed to set job mode: {}", e);
-                }
-            });
+                    .map_err(|e| {
+                        ToolError::ExecutionFailed(format!("failed to persist job mode: {}", e))
+                    })?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/tools/builtin/job.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/history/store/conversations/singletons.rs [hunk 2]
```diff
@@ -130,11 +130,12 @@ mod tests {
...
-    async fn cleanup_user(store: &Store, user_id: &str) {
-        let conn = store.conn().await.expect("connection should be available");
+    async fn cleanup_user(store: &Store, user_id: &str) -> Result<(), DatabaseError> {
+        let conn = store.conn().await?;
         conn.execute("DELETE FROM conversations WHERE user_id = $1", &[&user_id])
             .await
-            .expect("conversations should be deleted");
+            .map_err(DatabaseError::Postgres)?;
+        Ok(())
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/history/store/conversations/singletons.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

Please address the following concerns and ensure all commit gates succeed:

### ⚠️ Outside diff range comments (4)

#### src/db/libsql/routines/mapping.rs (1)

`92-181`: _⚠️ Potential issue_ | _🟠 Major_
**Add regression tests for malformed and NULL numeric decoding now.**
Add unit tests and behavioural tests that assert:
- `dedup_window_secs = NULL` maps to `None`; malformed value returns `DatabaseError::Serialization`.
- `tokens_used = NULL` maps to `None`; malformed or out-of-range value returns `DatabaseError::Serialization`.

Lock these expectations to prevent fallback-to-zero regressions.
As per coding guidelines "All new functionality or changes must be guarded by unit tests" and "All new functionality or changes must be guarded by behavioural tests where applicable".

##### 🤖 Prompt for AI Agents

#### src/db/libsql/conversations/messages.rs (1)

`77-83`: _⚠️ Potential issue_ | _🟡 Minor_
**Use consistent error variant for pagination limit overflow.**
Line 78 uses `DatabaseError::Validation` for the `checked_add` overflow, but line 82 uses `DatabaseError::Query` for the `i64` conversion overflow. Both are input validation failures for the same parameter; use `Validation` consistently.

##### ♻️ Proposed fix

```diff
let fetch_limit_i64 = i64::try_from(fetch_limit).map_err(|_| {
-        DatabaseError::Query("conversation message pagination limit overflow".to_string())
+        DatabaseError::Validation("conversation message pagination limit overflow".to_string())
     })?;
```

##### 🤖 Prompt for AI Agents

#### src/tools/builtin/job.rs (1)

`321-321`: _⚠️ Potential issue_ | _🟡 Minor_
**Replace `.expect()` with proper error propagation.**
Line 321 uses `.expect("sandbox deps required")` in production code. As per coding guidelines, `.expect()` and `.unwrap()` are forbidden outside of tests; errors must be propagated with `Result` and `?`.

##### 🛡️ Proposed fix

```diff
-        let jm = self.job_manager.as_ref().expect("sandbox deps required");
+        let jm = self.job_manager.as_ref().ok_or_else(|| {
+            ToolError::ExecutionFailed("sandbox deps required but not configured".to_string())
+        })?;
```

##### 🤖 Prompt for AI Agents

#### src/db/libsql_migrations/v16_context.rs (1)

`134-145`: _⚠️ Potential issue_ | _🟡 Minor_
**Use the `version` parameter in the rollback warning instead of hardcoding "V16".**
The function accepts `version` as a parameter but the warning message hardcodes "V16". This will be misleading if the helper is reused for other migrations.

##### Proposed fix

```diff
if let Err(cleanup_error) = conn.execute_batch("ROLLBACK;").await {
                 tracing::warn!(
                     version,
                     name,
                     error = %cleanup_error,
-                    "libSQL V16 rollback failed"
+                    "libSQL migration rollback failed"
                 );
             }
```

##### 🤖 Prompt for AI Agents

### ♻️ Duplicate comments (1)

#### src/channels/web/handlers/job_control/restart.rs (1)

`155-162`: _🧹 Nitpick_ | _🔵 Trivial_
**Document the Worker-default invariant and consider atomicity.**
The two-step persistence (`save_sandbox_job` then `update_sandbox_job_mode`) lacks transactional atomicity. If the first succeeds but the second fails, the job record exists with an implicit Worker mode (NULL). This is **safe** because NULL defaults to Worker on load, but the invariant should be documented here.
Add an inline comment explaining: "Worker is the default mode; only non-default modes (ClaudeCode) are persisted to the database. NULL in job_mode implies Worker mode."

##### 📝 Proposed documentation

```diff
+    // Worker is the default mode; only non-default modes (ClaudeCode) are
+    // persisted to the database. NULL in job_mode implies Worker mode.
+    // The two writes are not atomic, but failure of update_sandbox_job_mode
+    // leaves the job in Worker mode, which is a safe fallback.
     let is_claude_code = mode == crate::orchestrator::job_manager::JobMode::ClaudeCode;
     if is_claude_code {
```

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- In `DefaultSettingsPersistence::save_default_settings`, upserting each key in a loop without a transaction can leave the default user’s settings partially updated on error; consider wrapping the per-key writes in a single transaction (or a backend API that batches these upserts) so the operation is atomic.
- The new `limit == 0` validation in `list_conversation_messages_paginated` returns a `DatabaseError::Query`, which conflates client-side validation failures with actual query errors; consider using a more specific error variant (or adding one) to distinguish invalid input from database issues.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[src/db/libsql/routines/mapping.rs](https://github.com/leynos/axinite/pull/99/files/10b84abf0de3081586e1cde748d5adfa83c2f4ce#diff-6ff29b67fd1ced626c2aefd3afdedff50c3448b7b7ffaf231547394510512aec)

Comment on file

~~~rust
    async fn test_dedup_window_null_yields_none() {
        let row = mock_routine_row_with_dedup_window(None, None)
            .await
            .expect("failed to create mock row");
        let routine = row_to_routine_libsql(&row).expect("failed to map row to routine");
        assert!(routine.guardrails.dedup_window.is_none());
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=99&biomarker=Code+Duplication&filename=src%2Fdb%2Flibsql%2Froutines%2Fmapping.rs&method=)
The module contains 2 functions with similar structure: tests.test_dedup_window_null_yields_none,tests.test_dedup_window_valid_value

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

[src/tools/builtin/job.rs](https://github.com/leynos/axinite/pull/99/files/10b84abf0de3081586e1cde748d5adfa83c2f4ce#diff-27112c41d861e733b6c1ff31aca71e0b14ed0057d8193f59c46174ee6bd77a6b)

Comment on file

~~~rust
    }

    /// Update sandbox job status in DB (fire-and-forget).
    /// Update sandbox job status in DB (fire-and-forget for best-effort updates).
~~~

❌ New issue: [Lines of Code in a Single File](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=99&biomarker=Lines+of+Code+in+a+Single+File&filename=src%2Ftools%2Fbuiltin%2Fjob.rs&method=)
This module has 1088 lines of code, improve code health by reducing it to 1000

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/db/libsql/routines/mapping.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: tests.test_dedup_window_null_yields_none,tests.test_dedup_window_valid_value

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/db/libsql/routines/mapping.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: tests.test_dedup_window_null_yields_none,tests.test_dedup_window_valid_value

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

### Issue 1 — `save_default_settings` atomicity — ⚠️ Documented but not fixed

The per-key write loop in `src/setup/persistence.rs` (Lines 40–44) is unchanged — there is still no transaction wrapping and no batch API. The partial-commit risk remains.

However, a doc comment (Lines 34–37) has been added that explicitly acknowledges this:

> "Note: Writes are incremental and partial commits may occur if a later `set_setting` fails. Callers that require atomicity should use a backend-level batch/transaction method instead."

The original suggestion was advisory ("consider wrapping"), so documenting the known limitation rather than restructuring is a defensible choice. The risk is real but is now visible to callers. If atomicity is required in a future scenario (e.g. applying many settings in a critical setup path), the doc comment already points to the correct remedy.

**Verdict:** Not structurally resolved, but the non-atomic behaviour is explicitly documented. No further action is required in this PR unless atomicity is a hard requirement — if so, a follow-up issue would be appropriate.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

### ❌ Failed checks (3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | Multiple critical test gaps and unresolved review comments remain unaddressed across snapshot handling, assertion strength, validation message consistency, pagination edge cases, and credential-grant normalisation. | Delete transient `.pending-snap` entries, replace weak assertions with explicit key-presence checks, align validation messages, add pagination limit unit tests (limit==0, overflow), and add regression test for malformed legacy description JSON normalisation. |
| User-Facing Documentation | ⚠️ Warning | User-facing API behaviour changed (pagination validation now returns 400 BAD\_REQUEST) but no documentation was updated to reflect this change. | Update user-facing documentation to document the new pagination validation behaviour and the corresponding error message for limit <= 0. |
| Developer Documentation | ⚠️ Warning | PR introduces new internal APIs and architectural changes without documentation updates to developers-guide.md or database-integrations.md. | Update database-integrations.md to document DatabaseError::Validation distinction and migration helper patterns. Update developers-guide.md to explain job persistence patterns and parameter-struct conventions with code examples. |
